src/Lib.hs

```diff
@@ -59,3 +59,16 @@ checkShadowing args (Abs arg expr)
       nested  = checkShadowing (arg : args) expr
 checkShadowing args (App t u) = concatMap (checkShadowing args) [t, u]
 checkShadowing _ _ = []
+
+etaReduce :: Expr -> Expr
+etaReduce expr = case expr of
```

Comment:
The `case of` statement can be replaced with pattern matching on the function definition level

Reply:
I just wasn't sure how to access initial `expr` when using pattern matching on the function definition level. Is there a way to return inital argument without copy-pasting `(Abs ident (App expr' (Term ident')))`?

Reply:
Yes, you can use @-syntax: `etaReduce expr@(Abs ident (...))`

Reply:
Cool, thanks!